src/search/task_proxy.h

```diff
@@ -729,13 +718,90 @@ inline bool does_fire(const EffectProxy &effect, const State &state) {
     return true;
 }
 
-inline bool does_fire(const EffectProxy &effect, const GlobalState &state) {
-    for (FactProxy condition : effect.get_conditions()) {
-        FactPair condition_pair = condition.get_pair();
-        if (state[condition_pair.var] != condition_pair.value)
-            return false;
+inline bool State::operator==(const State &other) const {
+    assert(task == other.task);
+    if (registry != other.registry) {
+        std::cerr << "Comparing registered states with unregistered states "
+                  << "or registered states from different registries is "
+                  << "treated as an error because it is likely not "
+                  << "intentional."
+                  << std::endl;
+        utils::exit_with(utils::ExitCode::SEARCH_CRITICAL_ERROR);
+    }
+    if (registry) {
+        // Both states are registered and from the same registry.
+        return id == other.id;
+    }
+    // Both states are unregistered.
+    assert(values);
+    assert(other.values);
+    assert(!values->empty());
+    return *values == *other.values;
+}
+
+inline bool State::operator!=(const State &other) const {
+    return !(*this == other);
+}
+
+inline void State::unpack() const {
```

Comment:
Perhaps we should explain the reasoning for the `mutable` and `const` stuff somewhere, ideally where we declare the unpacked data to be mutable along the lines of:

unpacked_data is mutable because we think of it as a redundant representation of the state's contents, a kind of cache. One could argue for doing this differently (for example because some methods *require* unpacked data, so in some sense its presence changes the "logical state" from the program perspective. It is a bit weird to have a const function like unpack that is only called for its side effect on the object. But we decided to use `const` here to mean "const from the perspective of the state space semantics of the state".